rts/motoko-rts/src/types.rs

```diff
@@ -262,57 +263,111 @@ impl Value {
         unskew(self.0 as usize)
     }
 
-    /// Get the object tag. In debug mode panics if the value is not a pointer.
+    /// Check that the forwarding pointer is valid.
+    #[inline]
+    pub unsafe fn check_forwarding_pointer(self) {
+        debug_assert!(self.forward().get_ptr() == self.get_ptr());
+    }
+
+    /// Check whether the object's forwarding pointer refers to a different location.
+    pub unsafe fn is_forwarded(self) -> bool {
+        self.check_forwarding_pointer();
+        self.forward().get_ptr() != self.get_ptr()
+    }
+
+    /// Get the object tag. No forwarding. Can be applied to any block, regular objects
+    /// with a header as well as `OneWordFiller`, `FwdPtr`, and `FreeSpace`.
+    /// In debug mode panics if the value is not a pointer.
     pub unsafe fn tag(self) -> Tag {
-        debug_assert!(self.get().is_ptr());
-        (self.get_ptr() as *mut Obj).tag()
+        *(self.get_ptr() as *const Tag)
+    }
+
+    /// Get the forwarding pointer. Used by the incremental GC.
+    pub unsafe fn forward(self) -> Value {
+        debug_assert!(self.is_obj());
+        debug_assert!(self.get_ptr() as *const Obj != null());
+        let obj = self.get_ptr() as *const Obj;
+        (*obj).forward
     }
 
-    /// Get the pointer as `Obj`. In debug mode panics if the value is not a pointer.
+    /// Resolve forwarding if the value is a pointer. Otherwise, return the same value.
+    pub unsafe fn forward_if_possible(self) -> Value {
+        if self.is_ptr() && self.get_ptr() as *const Obj != null() {
+            // Ignore null pointers used in text_iter.
```

Comment:
Why does text_iter use `null`? I wonder if this edge case could lead to bugs elsewhere.

Reply:
I assume it is an exception and could be improved by using a reference to the `Null` object. Maybe we should do this is in a separate PR? (Currently, I do not see a bug implication elsewhere but I had to consider this special case during GC work.)

Reply:
Sure no reason to do it here.